phy/reporting.nim

```diff
@@ -0,0 +1,25 @@
+## Implements the generic reporting architecture. Everything that wants to
+## report hints, warnings, or errors needs access to a ``ReportContext``-
+## derived type.
+
+import
+  vm/utils
+
+type
+  ReportContext*[T] = object of RootObj
+    ## The base type that all report context types must be derived from.
+    sendPrc*: proc(c: ref ReportContext[T], rep: sink T)
+      ## called when a message is sent to the context
+    fatalPrc*: proc(c: ref ReportContext[T], rep: sink T)
+      ## called when a fatal message is sent to the context. The procedure
+      ## must not return normally
+
+proc report*[T](c: ref ReportContext[T], rep: sink T) =
```

Comment:
Hm, I get what you mean.

My original thinking was that a caller of `report` doesn't know what will happen (nor does it need to), since the `ReportContext` implementation could promote warnings or hints to errors, but this does obfuscate the intention of the caller, making it harder to search for where errors/warnings/hints are reported.

Having said that, I'm starting to think that to-error promotion shouldn't be the responsibility of the `ReportContext`, but rather that of the reporter. What warning should be considered an error is something that's independent of how the error is handled.

Reply:
> Having said that, I'm starting to think that to-error promotion shouldn't be the responsibility of the `ReportContext`, but rather that of the reporter. What warning should be considered an error is something that's independent of how the error is handled.

It's an interesting question, by promoting a hint or warning to error eagerly, we're causing them to impact control flow of analysis, where they otherwise wouldn't. Doing so later will only impact output/formatting and control whether to cross key thresholds like proceeding to final code generation. The early exit is tempting as we save work, but analysis knowing its control flow has advantages too -- fewer partial analysis scenarios to reason about. Note we always get some awkward partial analysis scenarios where things like LSP or checking/linting have higher error maxes.

I'm not entirely sure where promotion should take place, it could well be eager, but in my mind the hint that the explicit terms `error/warn/hint/etc` provide are useful in setting up expectations of default behaviour.

Another option I see is data driven. If we separate "diagnostic code" vs "diagnostic message", where one "diagnostic code", something like "error identifier expected", could have multiple variations of messages associated with it in order to make the output more programmer friendly. Here we could use a naming convention for the diagnostic codes like `errDiagFoo` or `hintDiagBar`, where the name has a severity prefix, prior to promotion. As long as the naming is consistent, then finding instances in the code base should be easier.

Regex: `.+(\s|\()errDiag(.+)`, that should find any call, or command syntax, assuming we have that, in one go.

Reply:
> If we separate "diagnostic code" vs "diagnostic message"

I think diagnostic codes are a necessity -- I'm not sure how things like report configuration (e.g., which hints and warnings are enabled, which warnings are promoted to errors, etc.) could work otherwise.

Reply:
> Side note: my idea with `report` (now `error`) was/is that it doesn't guarantee an early exit (or any special control-flow, for that matter), ...

Sorry, old habit with NimSkull, thinking `error` changed control flow, yup that would be `fatal`.